### ISSUE TITLE
Instructions on how to use sbt-ci-release with gpg 2

### ISSUE THREAD

**Issue by mwz** [olafurpg/sbt-ci-release#95]
Hey Olaf, I had some issues with trying to get your plugin to work and publish a signed release with gpg2, so I think it would be a good idea to document this process in case anyone else comes across this issue; here is the problem I run into and what I had to do to fix it. 

1. I didn't use your setup-gpg action, because I discovered it uses gpg 1.4 and I wanted to use the latest version (is there any specific reason you're using a legacy version? Have you considered upgrading it?). I'm using an ubuntu runner, so I just installed gpg via apt: `sudo apt-get -y install gnupg2`.

1. Publishing snapshots worked fine (I'm assuming snapshot artifacts are actually published unsigned?).

1. When trying to publish a release I was running into the following when running the `ci-release` task:

	1. First I would get the following warnings when the key was imported:
		> gpg: key XXX: public key "XXX" imported
		> gpg: key XXX: error sending to agent: Inappropriate ioctl for device
		> gpg: error building skey array: Inappropriate ioctl for device

	1. Next, sbt-pgp would fail while signing the artifacts with the following error:
		> [error] gpg: no default secret key: No secret key
		> [error] gpg: signing failed: No secret key
		> [error] java.lang.RuntimeException: Failure running 'gpg --batch --passphrase *** --detach-sign --armor --use-agent --output xxx'.  Exit code: 2
		> [error] 	at scala.sys.package$.error(package.scala:30)
		> [error] 	at com.jsuereth.sbtpgp.CommandLineGpgSigner.sign(PgpSigner.scala:42)

	1. Obviously I double checked all the env variables to make sure those were correct and after a bit of digging, I got it to work by using the gpg agent and setting the `pinentry-mode loopback` option. As I discovered, gpg 2.1 ignores passphrases passed in via env vars or standard input, so using the loopback feature to tell the agent to ask the invoking program for the passphrase did the trick - here's how I configured gpg:
		```bash
		sudo apt-get update && sudo apt-get -y install gnupg2
		mkdir ~/.gnupg && chmod 700 ~/.gnupg
		echo use-agent >> ~/.gnupg/gpg.conf
		echo pinentry-mode loopback >> ~/.gnupg/gpg.conf
		echo allow-loopback-pinentry >> ~/.gnupg/gpg-agent.conf
		chmod 600 ~/.gnupg/*
		echo RELOADAGENT | gpg-connect-agent
		echo $PGP_SECRET | base64 --decode | gpg --import --no-tty --batch --yes
		```

I'm happy to open a PR and add this somewhere to the FAQ maybe if you also feel like this could be useful for others - It would have certainly made my life a bit easier if these instructions on how to make this work with gpg2 were in the readme.

Thanks for the awesome plugin!

**Comment by olafurpg** [comment]
Thank you for reporting! I would LOVE to get a contribution updating the docs to explain how to use gpg2. I spent a lot of time trying to figure out why it doesn’t work and gave up.

It would be best if the plugin can automatically detect when gpg2 is installed and setup the correct configuration. I try to avoid installation steps that require adding bash scripts to the repo.

**Comment by soronpo** [comment]
@mwz Just ran into this issue myself. Where did you place this extra code?

**Comment by olafurpg** [comment]
BTW this plugin doesn’t do anything special with gpg except calling out to “publishSigned”

**Comment by mwz** [comment]
@soronpo, you can find a working pipeline with gpg2 [here](https://github.com/beautiful-scala/scalastyle/blob/master/.github/workflows/release.yml#L16).

**Comment by olafurpg** [comment]
@mwz the reason I'm OK with using v1.4 is because it looks like it's still officially supported with no advertized "end of life", see 

<img width="359" alt="Screenshot 2020-01-19 at 17 20 33" src="https://user-images.githubusercontent.com/1408093/72685234-1e150200-3ae0-11ea-8eeb-6e75d1d2512b.png">

https://gnupg.org/download/index.html

I still think it's a great idea to support v2. Do you have any idea how incorporate your changes in https://github.com/beautiful-scala/scalastyle/blob/master/.github/workflows/release.yml#L16 into the this plugin so everybody can enjoy the fix?

**Comment by ScalaFanatic** [comment]
Does it make more sense for the fix/documentation to be introduced in [sbt-pgp](https://github.com/sbt/sbt-pgp), since that is the plugin that actually runs the `gpg` command? There seems to be a fundamental issue with providing the passphrase as a command argument instead of, for example, using `--passphrase-fd 0` (which seems to be the recommended approach).

At any rate, I tested this solution out and I discovered (the hard way) that you need to be using `dist: bionic` instead of `dist: xenial` in your `.travis.yml` in order for this to work - otherwise `gpg` will complain that `pinentry-mode loopback` is an "invalid option". I did not test it using other `dist`s.

I also found that you do not need this command:

```bash
sudo apt-get update && sudo apt-get -y install gnupg2
```

if you are using `dist: bionic`, as you can see in greenhat's merge above.

Anecdotally, I found that making these changes will fix other errors such as:

```
[error] gpg: gpg-agent is not available in this session
[error] java.lang.RuntimeException: Failure running 'gpg --batch --passphrase [secure] --detach-sign --armor --use-agent --output ...  Exit code: 2
[error] 	at scala.sys.package$.error(package.scala:30)
[error] 	at com.jsuereth.sbtpgp.CommandLineGpgSigner.sign(PgpSigner.scala:48)
[error] 	at com.jsuereth.sbtpgp.PgpSettings$.$anonfun$signingSettings$2(PgpSettings.scala:151)
[error] 	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:245)
[error] 	at scala.collection.immutable.Map$Map4.foreach(Map.scala:236)
[error] 	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:245)
[error] 	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:242)
[error] 	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
[error] 	at com.jsuereth.sbtpgp.PgpSettings$.$anonfun$signingSettings$1(PgpSettings.scala:146)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:67)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:281)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:19)
[error] 	at sbt.Execute.work(Execute.scala:290)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:281)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:178)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:37)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
[error] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[error] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[error] 	at java.base/java.lang.Thread.run(Thread.java:834)
```

I was experiencing this error whenever I tried to do a tag-triggered release (pull request-triggered releases of SNAPSHOTs were working fine). **I am still unsure whether this was fixed because I switched to `dist: bionic` or if it was because I switched to gpg 2.** I will not be bothering to find out, just sharing for reference.

Lastly, if you want to avoid this error in your Travis logs:

```
The command "mkdir ~/.gnupg && chmod 700 ~/.gnupg" exited with 1.
```

I would make this change:
```bash
# Before
mkdir ~/.gnupg && chmod 700 ~/.gnupg
# After
mkdir -p ~/.gnupg && chmod 700 ~/.gnupg
```

**Comment by ghostdogpr** [comment]
I just ran into this issue using the pre-made `openjdk:8-jdk-node` docker image on CircleCI. Thanks a lot @mwz for posting the fix, saved me a lot of efforts!

**Comment by DavidGregory084** [comment]
Thanks @mwz! I also ran into this while porting https://github.com/DavidGregory084/sbt-tpolecat to Github Actions

**Comment by eed3si9n** [comment]
I think I cracked this nuts. There are two issues that gpg 2.2 is encountering:
- sbt-ci-release using `--import`
- sbt-pgp signing with `--passphrase`

I fixed the second issue in https://github.com/sbt/sbt-pgp/pull/184 last night and released sbt-pgp 2.1.1. The first issue can be worked around by running a Bash script:

```bash
#!/bin/bash
 
echo $PGP_SECRET | base64 --decode | gpg  --batch --import
```

(sbt-ci-release could potentially fix this using version detection like sbt-pgp?)

For full explanation check out https://eed3si9n.com/auto-publish-sbt-plugin-from-github-actions#gpg2
